internal/core/domain/trade_model.go

```diff
@@ -1,64 +1,299 @@
+		MessageID:  id,
+		ErrMessage: errMsg,
+		ErrCode:    pkgswap.ErrCode(errCode),
+	})
+	return id, msg
+}
+
+func (p swapParser) DeserializeRequest(msg []byte) (SwapRequest, error) {
+	s := &pbswap.SwapRequest{}
+	if err := proto.Unmarshal(msg, s); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (p swapParser) DeserializeAccept(msg []byte) (SwapAccept, error) {
+	s := &pbswap.SwapAccept{}
+	if err := proto.Unmarshal(msg, s); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (p swapParser) DeserializeComplete(msg []byte) (SwapComplete, error) {
+	s := &pbswap.SwapComplete{}
+	if err := proto.Unmarshal(msg, s); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (p swapParser) DeserializeFail(msg []byte) (SwapFail, error) {
+	s := &pbswap.SwapFail{}
+	if err := proto.Unmarshal(msg, s); err != nil {
+		return nil, err
 	}
+	return s, nil
+}
+
+// IPsetManager defines the required methods to override the extraction of the
+// txid and of the final transaction in hex format from the PSET one.
+// The default one comes from go-elements.
+type IPsetManager interface {
+	GetTxID(psetBase64 string) (string, error)
+	GetTxHex(psetBase64 string) (string, string, error)
+}
+
+type psetManager struct{}
+
+func (m psetManager) GetTxID(psetBase64 string) (string, error) {
+	return transactionutil.GetTxIdFromPset(psetBase64)
+}
+
+func (m psetManager) GetTxHex(psetBase64 string) (string, string, error) {
+	return wallet.FinalizeAndExtractTransaction(
+		wallet.FinalizeAndExtractTransactionOpts{PsetBase64: psetBase64},
+	)
+}
+
+var (
+	// SwapParser ...
+	SwapParser ISwapParser
+	// PsetManager ...
+	PsetManager IPsetManager
 )
 
-type Timestamp struct {
-	Request  uint64
-	Accept   uint64
-	Complete uint64
-	Expiry   uint64
+func init() {
+	SwapParser = swapParser{}
+	PsetManager = psetManager{}
+}
+
+// SwapRequest is the abstracted representation of a SwapRequest message.
+type SwapRequest interface {
+	GetId() string
+	GetAssetP() string
+	GetAmountP() uint64
+	GetAssetR() string
+	GetAmountR() uint64
+	GetTransaction() string
+	GetInputBlindingKey() map[string][]byte
+	GetOutputBlindingKey() map[string][]byte
 }
 
+// SwapAccept is the abstracted representation of a SwapAccept message.
+type SwapAccept interface {
+	GetId() string
+	GetRequestId() string
+	GetTransaction() string
+	GetInputBlindingKey() map[string][]byte
+	GetOutputBlindingKey() map[string][]byte
+}
+
+// SwapComplete is the abstracted representation of a SwapComplete message.
+type SwapComplete interface {
+	GetId() string
+	GetAcceptId() string
+	GetTransaction() string
+}
+
+// SwapFail is the abstracted representation of a SwapFail message.
+type SwapFail interface {
+	GetId() string
+	GetMessageId() string
+	GetFailureCode() uint32
+	GetFailureMessage() string
+}
+
+// Swap is the data structure that represents any of the above swaps.
 type Swap struct {
-	ID      string
-	Message []byte
+	ID        string
+	Message   []byte
+	Timestamp uint64
```

Comment:
what this timestamp refers to?

Reply:
before this, Timestamp was a struct for those of the swaps and the expiration... this moves the swap timestamps inside the Swap struct and 2 ad hoc fields for the expiration and the settlement.